### PR TITLE
Patterns Factory & Factory method

### PR DESCRIPTION
**Pull request by AntonTmsh**
4.3. Factory
4.4. Factory method

### CODE REVIEW

File: FactoryMethod/ValidationError.cs [hunk 1]
```diff
@@ -0,0 +1,25 @@
+﻿using System;
+using System.Collections.Generic;
+
+namespace Epam.NetMentoring.FactoryMethod
+{
+    public class ValidationError
```
Comment (fojman) [inline FactoryMethod/ValidationError.cs:3]: sorry but didn't get the idea of using static method and static member, could you please share your thoughts?
not sure but -  even if we have good reasons to do so - we need at least be able to clean up errors, i.e. - you processed one feed, then process another but in report you'll show both feed errors? am right in my suggestions?

File: FactoryMethod/Program.cs [hunk 2]
```diff
@@ -0,0 +1,37 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Epam.NetMentoring.FactoryMethod
+{
+    class Program
```
Comment (fojman) [inline FactoryMethod/Program.cs]: Access modifiers?

File: FactoryMethod/FeedManagerFactory.cs [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+        {
+            if (feedType == "DeltaOne")
+            {
+                return new DeltaOneFeedManager();
+            }
+            else if (feedType == "EM")
+            {
+                return new EMFeedManager();
+            }
+            throw new Exception("Unsupported feed type");
```
Comment (fojman) [inline FactoryMethod/FeedManagerFactory.cs]: could we consider a bit more reasonable exception type?
i.e. - ArgumentException() or InvalidaOperation() also nice to report what type exacltly not supported:
-  XXXException($"{feedType} unknow/not-supported") smth like this

File: Factory/BOFA_Filter.cs [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Epam.NetMentoring.Factory
+{
+    public class BOFA_Filter : IFilter
+    {
+        private readonly int bofaAmount = 70;
```
Comment (fojman) [inline Factory/BOFA_Filter.cs]: it's compile-time constant  - better to use `const int`

File: Factory/Barclays_Filter.cs [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Epam.NetMentoring.Factory
+{
+    public class Barclays_Filter : IFilter
+    {
+        private readonly string barclaysType = "Option";
```
Comment (fojman) [inline Factory/Barclays_Filter.cs]: see above, should be const

File: Factory/Program.cs [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+using System.Threading.Tasks;
+
+namespace Epam.NetMentoring.Factory
+{
+    class Program
+    {
+        static void Main(string[] args)
+        {
+            var trade = new List<Trade> { new Trade(12, "Future", "test") };
+            IFilterFactory factory;
```
Comment (fojman) [inline Factory/Program.cs]: we can use shortened write up:
`var factory = new FilterFactory();`

File: Factory/Program.cs [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+            factory = new FilterFactory();
+            IFilter filter = factory.CreateFilter("Connacord");
+            filter.Match(trade);
```
Comment (fojman) [inline Factory/Program.cs]: why don't we check result of Match?

File: Factory/Trade.cs [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Epam.NetMentoring.Factory
+{
+     public class Trade
+    {
+        //private int _amount;
```
Comment (fojman) [inline Factory/Trade.cs]: it's a good practice to avoid commented out code in prod

File: FactoryMethod/Account.cs [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+        public static Account GetAccount(int id)
+        {
+            if (!_accaunts.ContainsKey(id))
+            {
+                _accaunts.Add(id, new Account(id));
+            }
+            return _accaunts[id];
+        }
+    }
+    //public class AccountDB
```
Comment (fojman) [inline FactoryMethod/Account.cs]: remove commented out code

File: FactoryMethod/FeedItem.cs [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+        public string ValuationDate { get; set; }
+        public int CurrentPrice { get; set; }
+        public int SourceAccountId { get; set; }
+    }
+    public class DeltaOneFeed : FeedItem
+    {
+        public int Isin { get; set; }
+        public DateTime MaturityDate { get; set; }
+
+        public int FeedItemId { get => base.CounterpartyId + base.PrincipalId; }
```
Comment (fojman) [inline FactoryMethod/FeedItem.cs:20]: interesting write up, maybe you wanted to write:
`public int FeedItemId  get => base.CounterpartyId + base.PrincipalId;`

File: FactoryMethod/Program.cs [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+            foreach (List<FeedItem> feeds in listOfFeeds)
+            {
+                var feedmanager =  managerFactory.CreateFeedManager(CheckFeedType(feeds));
+                feedmanager.Process(feeds);
+            }
+        }
+
+        private static string CheckFeedType(List<FeedItem> feed)
+        {
+            var item = feed.First();
```
Comment (fojman) [inline FactoryMethod/Program.cs]: what it list empty? maybe - .FirstOrDefault()?

File: FactoryMethod/Program.cs [hunk 12]
```diff
@@ -0,0 +1,37 @@
...
+
+            if (item.GetType() == typeof(EmFeed))
+            {
+                return "EM";
```
Comment (fojman) [inline FactoryMethod/Program.cs]: Better to use constants instead of burned in raw strings in code

**Review by fojman — CHANGES_REQUESTED** [review]

File: FactoryMethod/Program.cs [hunk 13]
```diff
@@ -0,0 +1,37 @@
...
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Epam.NetMentoring.FactoryMethod
+{
+    class Program
+    {
+        static void Main(string[] args)
+        {
+            List<List<FeedItem>> listOfFeeds = new List<List<FeedItem>>();
```
Comment (fojman) [inline FactoryMethod/Program.cs]: Please at least one real test - now this code does nothing(list is empty)

File: FactoryMethod/Account.cs [hunk 14]
```diff
@@ -0,0 +1,32 @@
...
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Epam.NetMentoring.FactoryMethod
+{
+    public class Account
+    {
+        private static readonly Dictionary<int, Account> _accaunts = new Dictionary<int, Account>();
```
Comment (fojman) [inline FactoryMethod/Account.cs]: typo - _accaunts -> _accounts

File: Factory/Trade.cs (hunk 8 above)
Comment (fojman) [reply Factory/Trade.cs]: also, it looks like this class can be made "read-only"/immutable - which is good ideal(can discuss this)

File: Factory/Program.cs (hunk 6 above)
Comment (fojman) [reply Factory/Program.cs]: I would like to see more real-life example - please use some trade generator, real filtering, and filtering results, at list print something to console :)

File: Factory/Connacord_Filter.cs [hunk 15]
```diff
@@ -0,0 +1,19 @@
...
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Epam.NetMentoring.Factory
+{
+    public class Connacord_Filter : IFilter
+    {
+        private readonly int connacordAmountMin = 10;
```
Comment (fojman) [inline Factory/Connacord_Filter.cs]: const int?

File: Factory/Barclays_Filter.cs [hunk 16]
```diff
@@ -0,0 +1,19 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Epam.NetMentoring.Factory
+{
+    public class Barclays_Filter : IFilter
```
Comment (fojman) [inline Factory/Barclays_Filter.cs]: Please use consistent naming across - I see  CamelCase and some `__`- style

**Review by fojman — CHANGES_REQUESTED** [review]
Oveall looks good, pls just fix my comments!

File: FactoryMethod/ValidationError.cs (hunk 1 above)
Comment (AntonTmsh) [reply FactoryMethod/ValidationError.cs:3]: ok, fixed

File: FactoryMethod/Program.cs (hunk 2 above)
Comment (AntonTmsh) [reply FactoryMethod/Program.cs]: ok

File: FactoryMethod/FeedManagerFactory.cs (hunk 3 above)
Comment (AntonTmsh) [reply FactoryMethod/FeedManagerFactory.cs]: ok

File: Factory/BOFA_Filter.cs (hunk 4 above)
Comment (AntonTmsh) [reply Factory/BOFA_Filter.cs]: ok

File: Factory/Barclays_Filter.cs (hunk 5 above)
Comment (AntonTmsh) [reply Factory/Barclays_Filter.cs]: ok

File: Factory/Trade.cs (hunk 8 above)
Comment (AntonTmsh) [reply Factory/Trade.cs]: ok

File: FactoryMethod/Account.cs (hunk 9 above)
Comment (AntonTmsh) [reply FactoryMethod/Account.cs]: ok

File: FactoryMethod/FeedItem.cs (hunk 10 above)
Comment (AntonTmsh) [reply FactoryMethod/FeedItem.cs:20]: Hmm,, look like I can't write in this way, I think .net version issue

File: FactoryMethod/Program.cs (hunk 11 above)
Comment (AntonTmsh) [reply FactoryMethod/Program.cs]: ok

File: FactoryMethod/Program.cs (hunk 12 above)
Comment (AntonTmsh) [reply FactoryMethod/Program.cs]: ok

File: FactoryMethod/Account.cs (hunk 14 above)
Comment (AntonTmsh) [reply FactoryMethod/Account.cs]: fixed

File: Factory/Connacord_Filter.cs (hunk 15 above)
Comment (AntonTmsh) [reply Factory/Connacord_Filter.cs]: ok

File: Factory/Barclays_Filter.cs (hunk 16 above)
Comment (AntonTmsh) [reply Factory/Barclays_Filter.cs]: ok

File: FactoryMethod/Program.cs (hunk 13 above)
Comment (AntonTmsh) [reply FactoryMethod/Program.cs]: ok

**Pull request merged**